ffi/athcon/bindings/go/athcon.go

```diff
@@ -31,19 +32,48 @@ const (
 	_ = uint(C.sizeof_athcon_address - len(Address{}))
 )
 
-type Error int32
+type Error struct {
+	// athcon-compatible error code
+	Code int32
+	// underlying Go error for additional context
+	Err error
+}
 
 func (err Error) IsInternalError() bool {
-	return err < 0
+	return err.Code < 0
 }
 
+// Implement the Error method to return a string representation
 func (err Error) Error() string {
-	return C.GoString(C.athcon_status_code_to_string(C.enum_athcon_status_code(err)))
+	if err.Err != nil {
+		return fmt.Sprintf("%s: %v", C.GoString(C.athcon_status_code_to_string(C.enum_athcon_status_code(err.Code))), err.Err)
+	}
+	return C.GoString(C.athcon_status_code_to_string(C.enum_athcon_status_code(err.Code)))
 }
 
-const (
-	Failure = Error(C.ATHCON_FAILURE)
-	Revert  = Error(C.ATHCON_REVERT)
+var (
+	Failure                   = Error{Code: C.ATHCON_FAILURE}
+	Revert                    = Error{Code: C.ATHCON_REVERT}
+	OutOfGas                  = Error{Code: C.ATHCON_OUT_OF_GAS}
+	InvalidInstruction        = Error{Code: C.ATHCON_INVALID_INSTRUCTION}
+	UndefinedInstruction      = Error{Code: C.ATHCON_UNDEFINED_INSTRUCTION}
+	StackOverflow             = Error{Code: C.ATHCON_STACK_OVERFLOW}
+	StackUnderflow            = Error{Code: C.ATHCON_STACK_UNDERFLOW}
+	BadJumpDestination        = Error{Code: C.ATHCON_BAD_JUMP_DESTINATION}
+	InvalidMemoryAccess       = Error{Code: C.ATHCON_INVALID_MEMORY_ACCESS}
+	CallDepthExceeded         = Error{Code: C.ATHCON_CALL_DEPTH_EXCEEDED}
+	StaticModeViolation       = Error{Code: C.ATHCON_STATIC_MODE_VIOLATION}
+	PrecompileFailure         = Error{Code: C.ATHCON_PRECOMPILE_FAILURE}
```

Comment:
Does it make sense to expose all possible errors to the users of these bindings? I suspect some of them will never be used. For example, the host side would never return `UndefinedInstruction`.

Reply:
I don't see any harm in wrapping them all. In my mind that's sort of the purpose of the bindings, and it should be pretty unopinionated about what the client does with the exposed data types.

Reply:
The harm is that it leaks the internal details of the VM to the users. These errors are exported, and any change in them is technically a breaking change. IMO, the rule of thumb is to expose as little as possible.